WalletWasabi/Crypto/ZeroKnowledge/NonInteractive/FiatShamirTransform/Statement.cs

```diff
@@ -0,0 +1,71 @@
+using System.Linq;
+using System.Collections.Generic;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.NonInteractive.FiatShamirTransform
+{
+	public class Statement : IStatement
+	{
+		public Statement(LinearRelation.Statement statement)
+		{
+			LinearRelation = statement;
+		}
+
+		// Although in principle this could be an interface type, since only linear
+		// relations are needed for our proofs this is specialized to that type.
+		public LinearRelation.Statement LinearRelation { get; }
+
+		// Plug witness data into this statement. The enumerator should have exactly
+		// one vector, corresponding to the sizes of the equations in the base case
+		// of a linear relation and disjunction, and in the case of conjunction
+		// since we use conjunctive normal form, provide exactly one witness for
+		// each sub-statement.
+		public IKnowledge ToKnowledge(IEnumerable<ScalarVector> witnesses)
+		{
+			Guard.NotNullOrEmpty(nameof(witnesses), witnesses);
+			Guard.True(nameof(witnesses), witnesses.Count() == 1);
+			var witness = Guard.NotNull(nameof(witnesses), witnesses.First());
```

Comment:
```suggestion
			ScalarVector witness = Guard.NotNull(nameof(witnesses), witnesses.First());
```

Reply:
i think @lontivero was not in favor of using static types for declarations. is there a specific reason or just clarity?

if not i will skip this for now as i'm going to try playing around with refactoring IEnumerable out of everything and this probably won't survive anyway